Implement quota tracking options per ObjectStore.

jmchilton opened the pull request:
Redoing #10221 /  #10977 but building on #14044.

## Overview

#6552 implemented the ability for admins to assign job outputs to different object stores at runtime (this could take into account tool/workflow injected parameters or just be based on user, tool, destination, cluster state, etc..). But all the stored data would consume the same quota - regardless of the source selected.

This pull request allows different object stores or different groups of object stores to have different quotas or no quota at all. This enables uses cases such as sending job to cheaper data when a user's quota is getting near full or allowing admin to setup tool and/of workflow parameters to send job outputs higher quality, more redundant storage based on user selected options or user preferences.

This is a substantial step forward toward allowing scratch-space histories, while I suspect we want to implement some higher level convince functions and interface around that (per history preferences, object store preferences types) - I think that would all be based on these abstractions - abstractions that allow even more flexibility for admins who require it.

## Implementation

This adds the quota tag to XML/YAML object store declarations - that allow specifying a "quota source label" for each objectstore in a nested objectstore or disabling quota all together on objectstores.

The following quota block would assign all this storage to a quota source labelled with ``s3``.

```
        <backend id="dynamic_s3" type="disk" weight="0">
            <quota source="s3" />
            <files_dir path="${temp_directory}/files_dynamic_s3"/>
```

Whereas this would disable quota usage for this object store altogether. 

```
        <backend id="temp_disk" type="disk" weight="0">
            <quota enabled="false" />
            <files_dir path="${temp_directory}/files_cloud_scratch"/>
```

In order to implement this a new table/model has been added to track a user's usage per quota source label - namely ``UserQuotaSourceUsage``. Object stores that did not have a source label are still tracked using the User model's ``disk_usage`` attribute. I've updated all the scripts that recalculate user usage.

## UI + API

The quota dialog adds the option to pick a quota source label from those defined on the object stores, though this option only appears if quota source labels are configured.

<img width="1398" alt="Screen Shot 2020-09-28 at 8 45 33 PM" src="https://user-images.githubusercontent.com/216771/94499839-b5803800-01cb-11eb-8bfc-f033cad50917.png">

Likewise, by default the quota meter is unaffected but when multiple quota source labels are configured the meter becomes a link that shows the usage of each quota source.

<img width="727" alt="Screen Shot 2022-06-16 at 7 13 19 PM" src="https://user-images.githubusercontent.com/216771/174194628-a442728f-3dd2-44df-b8f5-c0fd36f7a4f8.png">


A new API ``/api/users/<user_id|current>/usage`` enables this.

## Abstractions for #4840

While this PR adds significant complexity related to recalculating a User's quota - it does reduce the duplication, adds tests (made more useful by having fewer paths through the quota recalculation code), and bring object store information into the calculation. I think this is all stuff that would be needed for #4840 and currently missing.

Part of this establishes a pattern for how to exclude certain datasets from usage calculation both when it is being added (included in #4840) and when re-calculdated (not included in #4840).

The API endpoints for disk usage across object stores and the UI entry point for displaying that information will hopefully both enable a more robust implementation of #4840.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jmchilton committed:
get_quota in SQL

jmchilton committed:
Rename test_selection to be more specific.

davelopez commented on client/src/components/ObjectStore/DescribeObjectStore.vue:
```diff
@@ -0,0 +1,70 @@
+            <b-spinner v-if="isLoadingUsage" />
+            <QuotaUsageBar v-else-if="quotaUsage" :quota-usage="quotaUsage" :embedded="true" />
+        </QuotaSourceUsageProvider>
+        <div v-else>Galaxy has no quota configured fo this object store.</div>
```
```suggestion
        <div v-else>Galaxy has no quota configured for this object store.</div>
```

davelopez commented on client/src/components/ObjectStore/DescribeObjectStore.vue:
```diff
@@ -0,0 +1,70 @@
+<template>
+    <div>
+        <div>
+            <span v-localize>{{ what }}</span>
```
Do you plan to change the *`what`* text when reusing this component somewhere else? Otherwise, we can remove this prop for simplicity and just put here *`This dataset is stored in`*.

jmchilton replied:
This extension point is used in the downstream objectstore selection work: https://github.com/galaxyproject/galaxy/pull/14073/commits/54a7b5e8cb1290daeb84dfbe263aaa546c2617f5#diff-2f4446e0f440cc2a7ab91620b66ec59a7009e5bf913bd8ffcb856c1488fc1103

davelopez commented on client/src/components/ObjectStore/ObjectStoreRestrictionSpan.vue:
```diff
@@ -0,0 +1,38 @@
+<template>
+    <span class="stored-how" v-b-tooltip.hover :title="title">{{ text }}</span>
```
This is just to fix the `eslint` warning about the [ordering of attributes](https://eslint.vuejs.org/rules/attributes-order.html).
```suggestion
    <span v-b-tooltip.hover class="stored-how" :title="title">{{ text }}</span>
```

davelopez commented on client/src/components/User/DiskUsage/Quota/QuotaUsageProvider.js:
```diff
@@ -8,17 +8,39 @@ import { QuotaUsage } from "./model";
 // https://github.com/galaxyproject/galaxy/pull/10977 is available
 
 /**
```
The TODO comment above can now be removed :tada:

davelopez commented on lib/galaxy/managers/users.py:
```diff
@@ -699,6 +705,25 @@ def add_serializers(self):
+            usage["quota_percent"] = self.user_manager.quota(user, quota_source_label=quota_source_label)
+            usage["quota"] = self.user_manager.quota(user, total=True, quota_source_label=quota_source_label)
+            usage["quota_bytes"] = self.user_manager.quota_bytes(user, quota_source_label=quota_source_label)
+            usage["nice_total_disk_usage"] = util.nice_size(usage["total_disk_usage"])
```
In general, I'm more in favor of returning always the size in bytes instead of *a nicely formatted* string, and whoever consumes this value can format it as they like. It gives the consumer more flexibility (maybe to calculate stuff with the value) and is one thing less to maintain and worry about in the backend, also if we provide both is kind of redundant. But not a big deal, since this is already like this in many API responses and I might be missing a historical or important point of having this *nice size* in the backend.

jmchilton replied:
I don't recall why I added these - likely because the UI worked differently at the time or maybe just for consistency with the old API but I'm happy to drop them. I'll do that on a rebase.

davelopez commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -155,6 +156,11 @@ class LegacyHistoryContentsIndexParams(Model):
     dataset_details: Optional[DatasetDetailsType]
     deleted: Optional[bool]
     visible: Optional[bool]
+    sharable: Optional[bool] = Field(
```
I'm a bit confused here, I thought that the legacy version of the endpoint would be deprecated and we want to add this new parameter to the *latest* version of the index endpoint in `HistoryContentsIndexParams`.

jmchilton replied:
Do you still think this is the case? The things called legacy seem to be the default way the API is used? You need to flag something special to take the non-legacy path.

davelopez replied:
Hmmm... :thinking: That's a good point :laughing: 
The UI is using the other `dev` versions of the endpoint, they seem to differ mainly in how content is filtered, so I guess it makes sense to add this to the default legacy path.

davelopez commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1118,6 +1126,13 @@ def __index_legacy(
+
+        object_store_ids = None
+        sharable = legacy_params.sharable
+        if sharable is not None:
```
See comment above, I think this should be part of the `dev` version of the endpoint but I might be wrong.

davelopez reviewed (commented):
This looks awesome!
Just some minor comments below from my limited point of view :sweat_smile:

jmchilton committed:
private objectstores & dataset.sharable

Setup abstractions to prevent sharing transient or private per-user objects in an objectstore.

jmchilton committed:
Implement quota tracking options per ObjectStore.

jmchilton committed:
Rebuild client typescript schema.

jmchilton committed:
unit test fix

jmchilton committed:
Lint fixes...

jmchilton committed:
Drop nice_total_disk_usage from usage APIs.

jmchilton committed:
Improved error message in api_asserts.
